inception/utils.py

```diff
@@ -0,0 +1,144 @@
+import logging
+import re
+from http import HTTPStatus
+
+from fastapi import HTTPException
+from torch.cuda import OutOfMemoryError
+
+from inception.config import settings
+from inception.metrics import ERROR_COUNT
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
+)
+logger = logging.getLogger(__name__)
+
+
+def clean_text_for_json(text: str) -> str | None:
+    """
+    Clean and prepare text for JSON encoding.
+    Handles special characters, line breaks, and other potential JSON issues.
+
+    :param text: Text to clean.
+    :return: Cleaned text or None if there was an error.
+    """
+
+    if not text:
+        return ""
+
+    try:
+        # Remove null bytes and other control characters except newlines and tabs
+        text = "".join(
+            char
+            for char in text
+            if char == "\n" or char == "\t" or (32 <= ord(char) < 127)
+        )
+
+        # Replace tabs with spaces
+        text = text.replace("\t", " ")
+
+        text = text.replace("\r\n", "\n").replace("\r", "\n")
+
+        # Remove spaces at the beginning and end of each line
+        text = "\n".join(line.strip() for line in text.split("\n"))
+
+        # Remove multiple consecutive empty lines
+        text = re.sub(r"\n{3,}", "\n\n", text)
+
+        # Remove leading/trailing whitespace
+        text = text.strip()
+        return text
+    except Exception as e:
+        raise ValueError(f"Error cleaning text: {str(e)}")
+
+
+def preprocess_text(text: str) -> str:
+    """
+    Preprocess text for embedding generation.
+    Includes cleaning and validation steps.
+    """
+    try:
+        cleaned_text = clean_text_for_json(text)
+        if not cleaned_text:
+            raise ValueError("Text is empty after cleaning.")
+
+        return cleaned_text
+    except Exception as e:
+        raise ValueError(f"Error preprocessing text: {str(e)}")
```

Comment:
It sort of looks like we'd never raise the error about empty text because the `Exception as e` would catch it and raise the next error, no? I might be wrong...

Reply:
Yeah, well, in this case, if the empty text error was raised, it would be appended to the next `ValueError`, resulting in a message like:

`Error preprocessing text: Text is empty after cleaning.`

However, thinking about it again, I believe it’s better to simply raise the `ValueError` for empty text, which will be handled in `handle_exception`. Other errors should propagate and be raised so that unexpected issues can be sent to Sentry for debugging. If needed, we can catch additional exceptions in the future, but for now, we should check them.

Refactored `preprocess_text` and `clean_text_for_json` accordingly.